src/constraints.jl

```diff
@@ -0,0 +1,559 @@
+            return model === JuMP.owner_model(cref)
+        end
+
+        # Get the ConstraintData object
+        function _constraint_data(cref::$RefType)
+            return gdp_data(JuMP.owner_model(cref)).$loc[JuMP.index(cref)]
+        end
+
+        @doc """
+            JuMP.name(cref::$($RefType))
+
+        Get a constraint's name attribute.
+        """
+        function JuMP.name(cref::$RefType)
+            return _constraint_data(cref).name
+        end
+
+        @doc """
+            JuMP.set_name(cref::$($RefType), name::String)
+
+        Set a constraint's name attribute.
+        """
+        function JuMP.set_name(cref::$RefType, name::String)
+            _constraint_data(cref).name = name
+            _set_ready_to_optimize(JuMP.owner_model(cref), false)
+            return
+        end
+
+        @doc """
+            JuMP.constraint_object(cref::$($RefType))
+
+        Return the underlying constraint data for the constraint 
+            referenced by `cref`.
+        """
+        function JuMP.constraint_object(cref::$RefType)
+            return _constraint_data(cref).constraint
+        end
+
+        # Extend Base methods
+        function Base.:(==)(cref1::$RefType, cref2::$RefType)
+            return cref1.model === cref2.model && cref1.index == cref2.index
+        end
+        Base.copy(cref::$RefType) = cref
+        @doc """
+            Base.getindex(map::JuMP.GenericReferenceMap, cref::$($RefType))
+
+        ...
+        """
+        function Base.getindex(map::JuMP.ReferenceMap, cref::$RefType)
+            $RefType(map.model, JuMP.index(cref))
+        end
+    end
+end
+
+# Extend delete
+"""
+    JuMP.delete(model::JuMP.Model, cref::DisjunctionRef)
+
+Delete a disjunction constraint from the `GDP model`.
+"""
+function JuMP.delete(model::JuMP.Model, cref::DisjunctionRef)
+    @assert JuMP.is_valid(model, cref) "Disjunctive constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _disjunctions(model)
+    #delete each logical variable
+    for lv_idx in dict[cidx].constraint.disjuncts
+        JuMP.delete(model, LogicalVariableRef(model, lv_idx))
+    end
+    #delete from gdp_data
+    delete!(dict, cidx)
+    #not ready to optimize
+    _set_ready_to_optimize(model, false)
+    return 
+end
+
+"""
+    JuMP.delete(model::JuMP.Model, cref::DisjunctConstraintRef)
+
+Delete a disjunct constraint from the `GDP model`.
+"""
+function JuMP.delete(model::JuMP.Model, cref::DisjunctConstraintRef)
+    @assert JuMP.is_valid(model, cref) "Disjunctive constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _disjunct_constraints(model)
+    #delete from gdp_data
+    delete!(dict, cidx)
+    #delete from constraint_to_indicator mapping
+    delete!(_constraint_to_indicator(model), cidx)
+    # TODO: delete reformulated constraint? This would require mapping cidx to reformulated constraint index
```

Comment:
I don't think we need to since it sets `_set_ready_to_optimize(model, false)`.

Reply:
Are you referring to deleting the reformulated constraint (TODO comment)? Or to deleting the `constraint_to_indicator` mapping?

Reply:
I am referring to the TODO. And yes it would make sense to not bother deleting the reformulated variables either.